pkg/kubelet/kubelet_node_status.go

```diff
@@ -123,34 +123,21 @@ func (kl *Kubelet) tryRegisterWithAPIServer(node *v1.Node) bool {
 		return false
 	}
 
-	if existingNode.Spec.ExternalID == node.Spec.ExternalID {
```

Comment:
removal of the field from the API in 1.11 means version-skewed kubelets (1.9 and 1.10) will fail this check (existingNode.Spec.ExternalID will be "").

we have to leave the field in place for two releases after this code is removed from the kubelet so old kubelets don't crash and burn.

Reply:
We can't drop fields without incrementing the API version:
https://kubernetes.io/docs/reference/using-api/deprecation-policy/
>Rule #1: API elements may only be removed by incrementing the version of the API group.

In general, I guess this was a security-related behavior change (kubernetes/community#911), which takes precedence over this part of the deprecation policy?
>As with whole REST resources, an individual field which was present in API v1 must exist and function until API v1 is removed.

Reply:
Element is ill defined here but my read is it refers to kinds, not fields. The context:

> Once an API element has been added to an API group at a particular version, it can not be removed from that version or have its behavior significantly changed, regardless of track.
> 
> Note: For historical reasons, there are 2 “monolithic” API groups - “core” (no group name) and “extensions”. Resources will incrementally be moved from these legacy API groups into more domain-specific API groups.

Aside we haven't dropped it yet: https://github.com/kubernetes/kubernetes/blob/d02cf08e27f640f09ebd489e094176fd075f3463/staging/src/k8s.io/api/core/v1/types.go#L3668

It just won't be used or filled in yet. This won't break openapi validation and the data that old clients send in the field will be stored and returned.

Reply:
This clause clearly captures individual fields:
>As with whole REST resources, an individual field which was present in API v1 must exist and function until API v1 is removed.

Reply:
https://kubernetes.io/docs/reference/using-api/deprecation-policy/#fields-of-rest-resources

Reply:
>> As with whole REST resources, an individual field which was present in API v1 must exist and function until API v1 is removed. Unlike whole resources, the v2 APIs may choose a different representation for the field, as long as it can be round-tripped. For example a v1 field named “magnitude” which was deprecated might be named “deprecatedMagnitude” in API v2. When v1 is eventually removed, the deprecated field can be removed from v2.
>
> gross

I agree. A field can just as easily be left at the same name in v2, documented as deprecated, and dropped once the non-deprecated versions (v1) are removed and round-tripping the field to those non-deprecated versions is no longer necessary.

Reply:
Yes, changing the name is ugly and we should probably typically just tag it as deprecated. I think the point is that it still has to work in v1, which is what this PR changes...

Reply:
Expectations about the behavior of the field.

I don't understand the full context of this case, but if a producer sets a field, and consumers can expect it to be set, then the deprecation policy sounds like the producer shouldn't stop setting it or do anything else that changes the behavior of the field consumers expect.